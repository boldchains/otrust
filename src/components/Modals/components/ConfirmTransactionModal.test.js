import { cleanup, fireEvent, render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';

import ConfirmTransactionModal from './ConfirmTransactionModal';
import {
  ThemeWrapper,
  ExchangeContextWrapper,
  ModalContextWrapper,
  renderWithContext,
} from '../../../utils/testing';

const testProps = {
  submitTrans: jest.fn(),
};

const testExchangeContext = {
  askAmount: new BigNumber(0),
  bidAmount: new BigNumber(0),
  bidDenom: 'strong',
  status: 'Not Approved',
  strong: 'ETH',
  weak: 'NOM',
};

const testModalContext = {
  handleModal: jest.fn(),
};

describe('Given the ConfirmTransactionModal component', () => {
  describe('when the component is rendered with bidDenom strong and askAmount, bidAmount are BigNumber', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithContext(ConfirmTransactionModal, testProps);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should not show all limitOptions', () => {
      const { queryByText } = renderWithContext(ConfirmTransactionModal, testProps);
      expect(queryByText('No limit')).not.toBeInTheDocument();
      expect(queryByText('1%')).not.toBeInTheDocument();
      expect(queryByText('2.5%')).not.toBeInTheDocument();
      expect(queryByText('5%')).not.toBeInTheDocument();
    });

    it('should show all gasOptions', () => {
      const { getByText } = renderWithContext(ConfirmTransactionModal, testProps);
      expect(getByText('0.000 (Standard)')).toBeInTheDocument();
      expect(getByText('0.000 (Fast)')).toBeInTheDocument();
      expect(getByText('0.000 (Instant)')).toBeInTheDocument();
    });

    describe('and user clicks on Slippage Limit', () => {
      it('should show Slippage Limit options with the first active defaul option', async () => {
        const { getByText } = renderWithContext(ConfirmTransactionModal, testProps);

        fireEvent.click(getByText('Slippage Limit'));

        expect(getByText('No limit')).toBeInTheDocument();
        expect(getByText('1%')).toBeInTheDocument();
        expect(getByText('2.5%')).toBeInTheDocument();
        expect(getByText('5%')).toBeInTheDocument();

        expect(getByText('No limit')).toHaveStyle('color: rgb(225, 223, 235)');
        expect(getByText('No limit')).toHaveStyle('background-color: rgb(48, 46, 61)');
        expect(getByText('1%')).toHaveStyle('color: rgb(152, 149, 166)');
        expect(getByText('1%')).toHaveStyle('background-color: transparent');
        expect(getByText('2.5%')).toHaveStyle('color: rgb(152, 149, 166)');
        expect(getByText('2.5%')).toHaveStyle('background-color: transparent');
        expect(getByText('5%')).toHaveStyle('color: rgb(152, 149, 166)');
        expect(getByText('5%')).toHaveStyle('background-color: transparent');
      });

      describe('and user clicks on 5% option of Slippage Limits', () => {
        it('should make it active and deactivate No limit option', async () => {
          const { getByText } = renderWithContext(ConfirmTransactionModal, testProps);
          fireEvent.click(getByText('Slippage Limit'));
          fireEvent.click(getByText('5%'));

          expect(getByText('No limit')).toHaveStyle('color: rgb(152, 149, 166)');
          expect(getByText('No limit')).toHaveStyle('background-color: transparent');
          expect(getByText('1%')).toHaveStyle('color: rgb(152, 149, 166)');
          expect(getByText('1%')).toHaveStyle('background-color: transparent');
          expect(getByText('2.5%')).toHaveStyle('color: rgb(152, 149, 166)');
          expect(getByText('2.5%')).toHaveStyle('background-color: transparent');
          expect(getByText('5%')).toHaveStyle('color: rgb(225, 223, 235)');
          expect(getByText('5%')).toHaveStyle('background-color: rgb(48, 46, 61)');
        });
      });
    });

    describe('and user clicks on Modal CloseIcon', () => {
      it('should call handleModal function from Modal Context', async () => {
        const { queryByTestId } = render(
          ExchangeContextWrapper(
            ModalContextWrapper(ThemeWrapper(ConfirmTransactionModal, testProps), testModalContext),
            testExchangeContext
          )
        );

        fireEvent.click(queryByTestId('confirm-modal-close-icon'));
        expect(testModalContext.handleModal).toHaveBeenCalled();
      });
    });

    describe('and user clicks on Secondary button', () => {
      it('should call handleModal function from Modal Context', async () => {
        const { queryByTestId } = render(
          ExchangeContextWrapper(
            ModalContextWrapper(ThemeWrapper(ConfirmTransactionModal, testProps), testModalContext),
            testExchangeContext
          )
        );

        fireEvent.click(queryByTestId('confirm-modal-secondary-button'));
        expect(testModalContext.handleModal).toHaveBeenCalled();
      });
    });

    describe('and user clicks on Modal PrimaryButton', () => {
      it('should call submitTrans function from props', async () => {
        const { queryByTestId } = renderWithContext(ConfirmTransactionModal, testProps);

        fireEvent.click(queryByTestId('confirm-modal-primary-button'));
        expect(testProps.submitTrans).toHaveBeenCalled();
      });
    });
  });

  describe('and bidDenom is weak', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        ExchangeContextWrapper(
          ModalContextWrapper(ThemeWrapper(ConfirmTransactionModal, testProps), testModalContext),
          { ...testExchangeContext, bidDenom: 'weak' }
        )
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  // test fails. Need modify format18 function or component checks for BigNumbers
  describe.skip('and bidDenom is weak and askAmount, bidAmount are NOT BigNumbers', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        ExchangeContextWrapper(
          ModalContextWrapper(ThemeWrapper(ConfirmTransactionModal, testProps), testModalContext),
          { ...testExchangeContext, bidDenom: 'weak', askAmount: 0, bidAmount: 0 }
        )
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});

/*
describe('Given the ConfirmTransactionModal  component', () => {
  describe('when the component is rendered', () => {
    let originalFetch;

    beforeEach(() => {
      originalFetch = global.fetch;
      global.fetch = jest.fn(() =>
        Promise.resolve({
          json: () =>
            Promise.resolve({
              data: {
                fast: 7000000000,
                rapid: 15000001459,
                slow: 4000000000,
                standard: 4000000000,
                timestamp: 1625307668406,
              },
            }),
        })
      );
    });

    afterEach(() => {
      global.fetch = originalFetch;
    });

    afterEach(cleanup);
  });
});
*/
