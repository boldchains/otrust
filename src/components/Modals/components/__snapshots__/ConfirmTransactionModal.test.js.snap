// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the ConfirmTransactionModal component and bidDenom is weak should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gtsrHT cttJaW"
  >
    <button
      class="sc-hKFxyN cLTPxI"
      data-testid="confirm-modal-close-icon"
    >
      <svg
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          opacity="0.5"
        >
          <g
            transform="translate(-.453 .555)"
          >
            <path
              class="cls-3"
              d="M8.762.384a1.311 1.311 0 011.981 1.707l-.127.147-8.378 8.378A1.311 1.311 0 01.257 8.909l.127-.147z"
              transform="translate(6.453 6.445)"
            />
            <path
              class="cls-3"
              d="M.384.384A1.311 1.311 0 012.09.257l.147.127 8.379 8.38a1.31 1.31 0 01-1.706 1.98l-.147-.127L.384 2.236a1.31 1.31 0 010-1.852z"
              transform="translate(6.453 6.445)"
            />
          </g>
        </g>
      </svg>
    </button>
    <main>
      <h3
        class="sc-jrsJWt bjHJTa"
      >
        Confirm Transaction
      </h3>
      <div
        class="sc-iCoGMd fhYZXm"
      >
        <div
          class="sc-gKAaRy cPQBIX"
        >
          You're receiving
        </div>
        ~  
        <sup>
          ETH
        </sup>
      </div>
      <div
        class="sc-iklJeh kJcUXB"
      >
        <span>
          Current Exchange Rate
        </span>
        <strong>
          1 NOM = NaN ETH
        </strong>
      </div>
      <div
        class="sc-iklJeh kJcUXB"
      >
        <span>
          You're sending
        </span>
        <strong>
          0.000000 NOM
        </strong>
      </div>
      <div
        class="sc-iklJeh kJcUXB"
      >
        <div>
          <span>
            Wallet
          </span>
          <div>
            <strong />
          </div>
        </div>
        <div
          class="sc-ciSkZP cgtoMT"
        >
          <svg
            viewBox="0 0 255 236.4"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.3 0l-99.5 73.9 18.4-43.6z"
              fill="#E2761B"
              stroke="#E2761B"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <g
              fill="#E4761B"
              stroke="#E4761B"
              stroke-linecap="round"
              stroke-linejoin="round"
            >
              <path
                d="M12.6 0l98.7 74.6-17.5-44.3zm193.9 171.3L180 211.9l56.7 15.6 16.3-55.3zm-204.4.9l16.2 55.3L75 211.9l-26.5-40.6z"
              />
              <path
                d="M71.8 102.7L56 126.6l56.3 2.5-2-60.5zm111.3 0l-39-34.8-1.3 61.2 56.2-2.5zM75 211.9l33.8-16.5-29.2-22.8zm71.1-16.5l33.9 16.5-4.7-39.3z"
              />
            </g>
            <path
              d="M180 211.9l-33.9-16.5 2.7 22.1-.3 9.3zm-105 0l31.5 14.9-.2-9.3 2.5-22.1z"
              fill="#D7C1B3"
              stroke="#D7C1B3"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M107 158l-28.2-8.3 19.9-9.1zm40.9 0l8.3-17.4 20 9.1z"
              fill="#233447"
              stroke="#233447"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M75 211.9l4.8-40.6-31.3.9zm100.2-40.6l4.8 40.6 26.5-39.7zm23.8-44.7l-56.2 2.5L148 158l8.3-17.4 20 9.1zM78.8 149.7l20-9.1L107 158l5.3-28.9-56.3-2.5z"
              fill="#CD6116"
              stroke="#CD6116"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M56 126.6l23.6 46-.8-22.9zm120.3 23.1l-1 22.9 23.7-46zm-64-20.6L107 158l6.6 34.1 1.5-44.9zm30.5 0l-2.7 18 1.2 45L148 158z"
              fill="#E4751F"
              stroke="#E4751F"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M148 158l-6.7 34.1 4.8 3.3 29.2-22.8 1-22.9zm-69.2-8.3l.8 22.9 29.2 22.8 4.8-3.3L107 158z"
              fill="#F6851B"
              stroke="#F6851B"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M148.5 226.8l.3-9.3-2.5-2.2h-37.7l-2.3 2.2.2 9.3L75 211.9l11 9 22.3 15.5h38.3l22.4-15.5 11-9z"
              fill="#C0AD9E"
              stroke="#C0AD9E"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M146.1 195.4l-4.8-3.3h-27.7l-4.8 3.3-2.5 22.1 2.3-2.2h37.7l2.5 2.2z"
              fill="#161616"
              stroke="#161616"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M246.5 78.7l8.5-40.8L242.3 0l-96.2 71.4 37 31.3 52.3 15.3 11.6-13.5-5-3.6 8-7.3-6.2-4.8 8-6.1zM0 37.9l8.5 40.8-5.4 4 8 6.1L5 93.6l8 7.3-5 3.6L19.5 118l52.3-15.3 37-31.3L12.6 0z"
              fill="#763D16"
              stroke="#763D16"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M235.4 118l-52.3-15.3 15.9 23.9-23.7 46 31.2-.4H253zM71.8 102.7L19.5 118 2.1 172.2h46.4l31.1.4-23.6-46zm71 26.4l3.3-57.7 15.2-41.1H93.8l15 41.1 3.5 57.7 1.2 18.2.1 44.8h27.7l.2-44.8z"
              fill="#F6851B"
              stroke="#F6851B"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </main>
    <section
      class="sc-jJMGnK JzxiD"
    >
      <p
        class="sc-hiKfDv VNUzn"
      >
        Gas Fee
      </p>
      <div
        class="sc-cBoqAE bSmxro"
      >
        <button
          class="sc-jcwpoC fufVtY"
        >
          0 (Standard)
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          0 (Fast)
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          0 (Instant)
        </button>
      </div>
      <p
        class="sc-hiKfDv VNUzn"
      >
        Slippage Limit
      </p>
      <div
        class="sc-cBoqAE bSmxro"
      >
        <button
          class="sc-jcwpoC ZvOvV"
        >
          No limit
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          1%
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          2.5%
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          5%
        </button>
      </div>
      <p
        class="sc-gXfVKN bRUKru"
      >
        Slippage is likely in times of high demand. Quote is based on most recent block and does not reflect transactions ahead of you in the mempool
      </p>
    </section>
    <footer>
      <div
        class="sc-dlnjwi excNNS"
      >
        <button
          class="sc-eCApnc sc-jSFjdj kclkZg hzEfXb"
          data-testid="confirm-modal-secondary-button"
        >
          Cancel
        </button>
        <button
          class="sc-eCApnc kclkZg"
          data-testid="confirm-modal-primary-button"
        >
          Confirm (60)
        </button>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;

exports[`Given the ConfirmTransactionModal component when the component is rendered with bidDenom strong and askAmount, bidAmount are BigNumber should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gtsrHT cttJaW"
  >
    <button
      class="sc-hKFxyN cLTPxI"
      data-testid="confirm-modal-close-icon"
    >
      <svg
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          opacity="0.5"
        >
          <g
            transform="translate(-.453 .555)"
          >
            <path
              class="cls-3"
              d="M8.762.384a1.311 1.311 0 011.981 1.707l-.127.147-8.378 8.378A1.311 1.311 0 01.257 8.909l.127-.147z"
              transform="translate(6.453 6.445)"
            />
            <path
              class="cls-3"
              d="M.384.384A1.311 1.311 0 012.09.257l.147.127 8.379 8.38a1.31 1.31 0 01-1.706 1.98l-.147-.127L.384 2.236a1.31 1.31 0 010-1.852z"
              transform="translate(6.453 6.445)"
            />
          </g>
        </g>
      </svg>
    </button>
    <main>
      <h3
        class="sc-jrsJWt bjHJTa"
      >
        Confirm Transaction
      </h3>
      <div
        class="sc-iCoGMd fhYZXm"
      >
        <div
          class="sc-gKAaRy cPQBIX"
        >
          You're receiving
        </div>
        ~  
        <sup>
          wNOM
        </sup>
      </div>
      <div
        class="sc-iklJeh kJcUXB"
      >
        <span>
          Current Exchange Rate
        </span>
        <strong>
          1 ETH = NaN NOM
        </strong>
      </div>
      <div
        class="sc-iklJeh kJcUXB"
      >
        <span>
          You're sending
        </span>
        <strong>
          0.000000 ETH
        </strong>
      </div>
      <div
        class="sc-iklJeh kJcUXB"
      >
        <div>
          <span>
            Wallet
          </span>
          <div>
            <strong />
          </div>
        </div>
        <div
          class="sc-ciSkZP cgtoMT"
        >
          <svg
            viewBox="0 0 255 236.4"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.3 0l-99.5 73.9 18.4-43.6z"
              fill="#E2761B"
              stroke="#E2761B"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <g
              fill="#E4761B"
              stroke="#E4761B"
              stroke-linecap="round"
              stroke-linejoin="round"
            >
              <path
                d="M12.6 0l98.7 74.6-17.5-44.3zm193.9 171.3L180 211.9l56.7 15.6 16.3-55.3zm-204.4.9l16.2 55.3L75 211.9l-26.5-40.6z"
              />
              <path
                d="M71.8 102.7L56 126.6l56.3 2.5-2-60.5zm111.3 0l-39-34.8-1.3 61.2 56.2-2.5zM75 211.9l33.8-16.5-29.2-22.8zm71.1-16.5l33.9 16.5-4.7-39.3z"
              />
            </g>
            <path
              d="M180 211.9l-33.9-16.5 2.7 22.1-.3 9.3zm-105 0l31.5 14.9-.2-9.3 2.5-22.1z"
              fill="#D7C1B3"
              stroke="#D7C1B3"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M107 158l-28.2-8.3 19.9-9.1zm40.9 0l8.3-17.4 20 9.1z"
              fill="#233447"
              stroke="#233447"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M75 211.9l4.8-40.6-31.3.9zm100.2-40.6l4.8 40.6 26.5-39.7zm23.8-44.7l-56.2 2.5L148 158l8.3-17.4 20 9.1zM78.8 149.7l20-9.1L107 158l5.3-28.9-56.3-2.5z"
              fill="#CD6116"
              stroke="#CD6116"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M56 126.6l23.6 46-.8-22.9zm120.3 23.1l-1 22.9 23.7-46zm-64-20.6L107 158l6.6 34.1 1.5-44.9zm30.5 0l-2.7 18 1.2 45L148 158z"
              fill="#E4751F"
              stroke="#E4751F"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M148 158l-6.7 34.1 4.8 3.3 29.2-22.8 1-22.9zm-69.2-8.3l.8 22.9 29.2 22.8 4.8-3.3L107 158z"
              fill="#F6851B"
              stroke="#F6851B"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M148.5 226.8l.3-9.3-2.5-2.2h-37.7l-2.3 2.2.2 9.3L75 211.9l11 9 22.3 15.5h38.3l22.4-15.5 11-9z"
              fill="#C0AD9E"
              stroke="#C0AD9E"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M146.1 195.4l-4.8-3.3h-27.7l-4.8 3.3-2.5 22.1 2.3-2.2h37.7l2.5 2.2z"
              fill="#161616"
              stroke="#161616"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M246.5 78.7l8.5-40.8L242.3 0l-96.2 71.4 37 31.3 52.3 15.3 11.6-13.5-5-3.6 8-7.3-6.2-4.8 8-6.1zM0 37.9l8.5 40.8-5.4 4 8 6.1L5 93.6l8 7.3-5 3.6L19.5 118l52.3-15.3 37-31.3L12.6 0z"
              fill="#763D16"
              stroke="#763D16"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M235.4 118l-52.3-15.3 15.9 23.9-23.7 46 31.2-.4H253zM71.8 102.7L19.5 118 2.1 172.2h46.4l31.1.4-23.6-46zm71 26.4l3.3-57.7 15.2-41.1H93.8l15 41.1 3.5 57.7 1.2 18.2.1 44.8h27.7l.2-44.8z"
              fill="#F6851B"
              stroke="#F6851B"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </main>
    <section
      class="sc-jJMGnK JzxiD"
    >
      <p
        class="sc-hiKfDv VNUzn"
      >
        Gas Fee
      </p>
      <div
        class="sc-cBoqAE bSmxro"
      >
        <button
          class="sc-jcwpoC fufVtY"
        >
          0 (Standard)
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          0 (Fast)
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          0 (Instant)
        </button>
      </div>
      <p
        class="sc-hiKfDv VNUzn"
      >
        Slippage Limit
      </p>
      <div
        class="sc-cBoqAE bSmxro"
      >
        <button
          class="sc-jcwpoC ZvOvV"
        >
          No limit
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          1%
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          2.5%
        </button>
        <button
          class="sc-jcwpoC fufVtY"
        >
          5%
        </button>
      </div>
      <p
        class="sc-gXfVKN bRUKru"
      >
        Slippage is likely in times of high demand. Quote is based on most recent block and does not reflect transactions ahead of you in the mempool
      </p>
    </section>
    <footer>
      <div
        class="sc-dlnjwi excNNS"
      >
        <button
          class="sc-eCApnc sc-jSFjdj kclkZg hzEfXb"
          data-testid="confirm-modal-secondary-button"
        >
          Cancel
        </button>
        <button
          class="sc-eCApnc kclkZg"
          data-testid="confirm-modal-primary-button"
        >
          Confirm (60)
        </button>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
